### ISSUE TITLE
Fix ability to run tests

### ISSUE THREAD

**Issue by MisinformedDNA** [dotnet/corefxlab#2224]
When I try to run tests in VS2017P3, I get this error

> Microsoft.VisualStudio.TestPlatform.ObjectModel.TestPlatformException: Testhost process exited with error: It was not possible to find any compatible framework version
> The specified framework 'Microsoft.NETCore.App', version '2.1.0-preview3-26410-06' was not found.
>   - Check application dependencies and target a framework version installed at:
>       C:\Program Files\dotnet\
>   - Installing .NET Core prerequisites might help resolve this problem:
>       http://go.microsoft.com/fwlink/?LinkID=798306&clcid=0x409
>   - The .NET Core framework and SDK can be installed from:
>       https://aka.ms/dotnet-download
>   - The following versions are installed:
>       1.0.1 at [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
>       1.0.4 at [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
>       1.0.5 at [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
>       1.1.1 at [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
>       1.1.2 at [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
>       2.0.0 at [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
>       2.0.3 at [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
>       2.0.5 at [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
>       2.0.6 at [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
>       2.1.0-preview1-26216-03 at [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
>       2.1.0-preview2-26406-04 at [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
> 
>    at Microsoft.VisualStudio.TestPlatform.CrossPlatEngine.Client.ProxyOperationManager.SetupChannel(IEnumerable`1 sources, CancellationToken cancellationToken)
>    at Microsoft.VisualStudio.TestPlatform.CrossPlatEngine.Client.ProxyExecutionManager.StartTestRun(TestRunCriteria testRunCriteria, ITestRunEventsHandler eventHandler)
 
OK, so it suddenly wants `2.1.0-preview3-26410-06`, but I don't have it installed. I'm guessing something inserted itself during the `build.cmd`.

From the command line, `dotnet --version` gives me `2.1.300-preview2-008533`, which doesn't match any of the above.

Also, if I try to test from the command line, using `build tests\System.IO.FileSystem.Watcher.Polling.Tests\System.IO.FileSystem.Watcher.Polling.Tests.csproj`, I get

> D:\Repos\corefxlab\corefxlab.sln.metaproj : error MSB4126: The specified solution configuration "tests\System.IO.FileSystem.Watcher.Polling.Tests\System.IO.FileSystem.Watcher.Polling.Tests.csproj|Any CPU" is invalid. Please specify a valid solution configuration using the Configuration and Platform properties (e.g. MSBuild.exe Solution.sln /p:Configuration=Debug /p:Platform="Any CPU") or leave those properties blank to use the default solution configuration. [D:\Repos\corefxlab\corefxlab.sln]

What do I need to do so I can test again?

**Comment by ahsonkhan** [comment]
If you got the latest master, we had package dependency updates (and will continue to get them): https://github.com/dotnet/corefxlab/pull/2221

Delete the dotnetcli directory and run build.cmd to install the updated version.

Like I said previously (in https://github.com/dotnet/corefxlab/issues/2219#issuecomment-381712911), we need to update our build script. PRs are welcome :)

**Comment by MisinformedDNA** [comment]
Ha! I knew I had asked a similar question somewhere.

`build` resulted in an error, so I did a `git clean -dXf` and it says

> D:\Repos\corefxlab>git clean -dXf
> Unlink of file 'dotnetcli/dotnet.exe' failed. Should I try again? (y/n) y
> Unlink of file 'dotnetcli/dotnet.exe' failed. Should I try again? (y/n) y
> Unlink of file 'dotnetcli/dotnet.exe' failed. Should I try again? (y/n)

So I went to Task Manager and found 7 dotnet orphan processes running. I'm not doing anything with them, so I shut them all down and delete succeeded. Are the orphaned processes a known issue?

And I'm happy to help with a PR. If you can point me in the right direction, I'll take a look. Obviously, I still haven't fully understood the process of working with this library, let alone everything going on "behind the scenes".

**Comment by ahsonkhan** [comment]
> So I went to Task Manager and found 7 dotnet orphan processes running. I'm not doing anything with them, so I shut them all down and delete succeeded. Are the orphaned processes a known issue?

We should also create a clean.cmd script (similar to corefx) that shuts down any of these orphaned processes and does a clean install of the correct dotnetcli so devs don't have to repeat these steps.

> And I'm happy to help with a PR. If you can point me in the right direction, I'll take a look. Obviously, I still haven't fully understood the process of working with this library, let alone everything going on "behind the scenes".

Start with https://github.com/dotnet/corefxlab/blob/master/scripts/build.ps1#L15
- Append a new check, something along the lines of, if the folder exists, but the cli version doesn't match what's in `DotnetCLIVersion.txt` or the runtime version doesn't match what's in `SharedRuntimeVersion.txt`, go ahead and install the correct cli with runtime (https://github.com/dotnet/corefxlab/blob/master/scripts/build.ps1#L26-L32).
   - This should fix the issue of having to delete the dotnetcli folder to get the latest package installed.
- Next, create a clean.cmd file that essentially does the cleanup for you and then calls does the install steps from the build.ps1.

**Comment by ahsonkhan** [comment]
> Obviously, I still haven't fully understood the process of working with this library, let alone everything going on "behind the scenes".

Please keep asking these questions! Better yet, document your process and update the README.md (or add a developer-guide) to include the steps necessary to unblock yourself. This is valuable feedback and I am sure other new contributors would have similar concerns.

**Comment by MisinformedDNA** [comment]
I created a PoC app in the solution for my personal use, but it's causing the build error:

> D:\Repos\corefxlab\corefxlab.sln.metaproj : error MSB4126: The specified solution configuration "ConsoleApp1\ConsoleApp1.csproj|Any CPU" is invalid. Please specify a valid solution configuration using the Configuration and Platform properties (e.g. MSBuild.exe Solution.sln /p:Configuration=Debug /p:Platform="Any CPU") or leave those properties blank to use the default solution configuration. [D:\Repos\corefxlab\corefxlab.sln]

Given that my project file looks similar to other project files in the repo, I'm not seeing where the issue is. Is there something special happening with a console app?

```xml
<Project Sdk="Microsoft.NET.Sdk">
  <Import Project="..\tools\common.props" />
  <PropertyGroup>
    <OutputType>Exe</OutputType>
    <TargetFramework>netcoreapp2.1</TargetFramework>
  </PropertyGroup>
  <ItemGroup>
    <ProjectReference Include="..\src\System.IO.FileSystem.Watcher.Polling\System.IO.FileSystem.Watcher.Polling.csproj" />
  </ItemGroup>
</Project>
```

**Comment by MisinformedDNA** [comment]
When `build.ps1` is called. By default `BuildVersion` is [set](https://github.com/dotnet/corefxlab/blob/master/scripts/build.ps1#L5) to `preview2-yyMMdd-1`. This version number is then [passed onto](https://github.com/dotnet/corefxlab/blob/master/scripts/build.ps1#L64) the solution build where it subsequently is used by `Microsoft.NET.DefaultAssemblyInfo.targets`. My question is: Do I care what this suffix is? Does it affect anything I'm doing?

The thing that stood out to me was that the suffix is set to `preview2` instead of `preview3`, but I just wanted to verify that this is meaningless to me, which I think I can.

**Comment by MisinformedDNA** [comment]
> the cli version doesn't match what's in DotnetCLIVersion.txt or the runtime version doesn't match what's in SharedRuntimeVersion.txt

I get the cli version with `dotnet --version`, correct? Now, how do I get the runtime version? Do I just want to make sure that the shared runtime is in the list from `dotnet --fx-version`? Or parse it out from "Host, Version"? Other?

**Comment by MisinformedDNA** [comment]
If I look at dotnet cli versions, I see that `DotnetCLIVersion.txt`'s version is before my cli version. (I'm not sure how I got in this state, but I'm guessing it has to do with different branches being at different points.)

```
D:\Repos\corefxlab>dotnet --version
2.1.300-preview2-008533

D:\Repos\corefxlab>more DotnetCLIVersion.txt
2.1.300-preview2-008530
```

Currently, the script tries to update to the `DotnetCLIVersion.txt`, but that version is already there, so it tries, realizes it's already there and moves on, but it will keep trying on each `build` since the versions continue to be off. Is this OK? I could do a "less-than" comparison, but I'm not sure how consistent the naming/numbering is.

> dotnet-install: .NET SDK version 2.1.300-preview2-008530 is already installed.

Thoughts?

**Comment by ahsonkhan** [comment]
> The thing that stood out to me was that the suffix is set to `preview2` instead of `preview3`, but I just wanted to verify that this is meaningless to me, which I think I can.

Since corefxlab is mainly for experimentation, we don't tend to follow the versioning of corefx/etc. too closely. We recently changed from `e-` to `preview2-` just as part of a specific SignalR release. 

> My question is: Do I care what this suffix is? Does it affect anything I'm doing?

No, this can be safely ignored and left untouched.

> I get the cli version with `dotnet --version`, correct? Now, how do I get the runtime version?

`dotnet --info` should give you the details.

> Currently, the script tries to update to the `DotnetCLIVersion.txt`, but that version is already there, so it tries, realizes it's already there and moves on, but it will keep trying on each `build` since the versions continue to be off. Is this OK? 

I don't think that is correct. There should be an exact match. Can you run `dotnet --info` and share the Base Path of where the dotnet.exe is actually coming from (assuming your corefxlab repo is in sync with master)? You may have locally installed a new dotnet cli (into Program Files\dotnet\...) and are possibly using that version instead.

If so, I would make sure to set the following environment variables (either in the developer window, or change the system settings):
`set DOTNET_MULTILEVEL_LOOKUP=0`

Also, set path so that `<path to repo>\corefxlab\dotnetcli\` is listed above (or before) `C:\Program Files\dotnet\`

For example:
`set path`
> ...;D:\GitHub\Fork\corefxlab\dotnetcli\;C:\Program Files\dotnet\;...

**Comment by ahsonkhan** [comment]
> I created a PoC app in the solution for my personal use, but it's causing the build error

I will have to know a bit more about your project/etc. (how it was added to the solution) to help with the error you are seeing. Can you point me to your repo fork/branch and I can take a look.

**Comment by YohDeadfall** [comment]
It's closed by #2226 @ahsonkhan.

@MisinformedDNA you can use keywords to automatically close an associated issue when a pull request gets merged. Visit [the help center](https://help.github.com/articles/closing-issues-using-keywords/) for more details.
